### PR TITLE
Iterative Phase Estimation Bugfix

### PR DESCRIPTION
**Pull request by shifubear**
### Summary

Bugfix for the issue mentioned here: [#9280](https://github.com/Qiskit/qiskit-terra/issues/9280).

### Details and comments

Fixed algorithm to build the correct circuit, and added tests for the situations where the old iterative phase estimation algorithm was failing.

### CODE REVIEW

**Comment by shifubear** [comment]
Hi @woodsp-ibm, do you have any pointers on what may be causing the CI to fail, and what I might need to fix? It looks like it's timing out after an hour on the docs test which I haven't really touched.

**Comment by Cryoris** [comment]
@shifubear this looks like a timeout not related to your PR, we'll just re-run the checks which should fix the problem.

@jlapeyre could you have a look at this fix, as you know the phase estimators best? 🙂

File: releasenotes/notes/iterative-phase-estimation-bugfix-0f10810ab93acfba.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+prelude: >
+    The iterative phase estimation class had a bug where the incorrect circuit was being generated, causing the algorithm to fail for even simple cases. 
+fixes:
+  - |
+    Fixed a bug where the iterative phase estimation algorithm was generating the wrong circuit, causing the algorithm to fail for simple cases. Refer to `#9280 https://github.com/Qiskit/qiskit-terra/issues/9280` for more details.
```
Comment (jlapeyre) [inline releasenotes/notes/iterative-phase-estimation-bugfix-0f10810ab93acfba.yaml:6]: This file is missing a newline. Also, you can remove the prelude. It has never been used for a feature or bug-fix PR. Currently, only for release notes.

**Review by jlapeyre — CHANGES_REQUESTED** [review]
This looks good to go (except for a problem with formatting). I can't see any way to substantially improve this.

File: releasenotes/notes/iterative-phase-estimation-bugfix-0f10810ab93acfba.yaml (hunk 1 above)
Comment (shifubear) [reply releasenotes/notes/iterative-phase-estimation-bugfix-0f10810ab93acfba.yaml]: Good to know! Just made the requested changes.

**Comment by woodsp-ibm** [comment]
@shifubear For the CLA, which is needed to be signed/agreed for any contribution to be accepted, github needs to recognize the account via the email in your commits - this would be from your git config and it seems it does not match any email your shifubear github account has. This message above has more info https://github.com/Qiskit/qiskit-terra/pull/9289#issuecomment-1352357987 - the simplest solution is to add whatever that email is to the account.

**Comment by woodsp-ibm** [comment]
@shifubear As I commented https://github.com/Qiskit/qiskit-terra/pull/9289#issuecomment-1370147695 the email in the commits does not match any github account - its a MacBook-Pro local email looking at the commits. (The name/email comes from the git config used when this was committed from the machine you used.) The CLA needs to be agreed for this to move forwards.

**Comment by shifubear** [comment]
@woodsp-ibm I've tried to add my email to my workspace and make a new commit, but it seems like that didn't do the trick... Is there a way to work around this and add the email to older commits? If not, maybe I can open a new PR with the same changes.

**Comment by woodsp-ibm** [comment]
@shifubear When you say "workspace" you mean to your "github account"? You followed the instructions in the CLA message above about adding the email, the one in your git config that was used for these commits, to your githib account, and it still does not work? This usually fixes things for people. You can of course re-do this in a new PR first though changing your local git config on the machine you are using to set the email in that to one that is associated with your github account.

**Comment by javabster** [comment]
hi @shifubear are you still working on this?

**Comment by shifubear** [comment]
Hi @javabster, it has been resolved in a new PR that is referenced in this thread so it is safe to close.

**Comment by ElePT** [comment]
In that case, I will close it, but let me know if there is any issue.

**Pull request closed without merging**